Mathlib/SetTheory/Surreal/Division.lean

```diff
@@ -0,0 +1,714 @@
+`x'` : a positive option of `x`, indexed by `i`
+`y` : the purported inverse of `x`, as defined by `x.inv'`
+`y'` : an option of `y`, indexed by `i'`
+`y''` : `(1 + (x' - x) * y') / x'` , the form of any option of `y` besides 0, indexed by `i''`
+
+## References
+
+* [Conway, *On numbers and games*][Conway2001]
+
+-/
+open SetTheory
+open PGame
+open Equiv
+
+variable {x : PGame} (x_num : x.Numeric)
+
+namespace Surreal.Division
+section Positive
+
+variable (x_pos : 0 < x)
+
+/-! ### A few random positivity lemmas -/
+
+/-- This applies to an arbitrary `LinearCommOrderedRing` and probably should go elsewhere. -/
+lemma inv_pos_of_pos {x y : Surreal} (h_inv: x * y = 1) (h : 0 < x) : 0 < y := by
+  apply (pos_iff_pos_of_mul_pos _).mp h
+  rw [h_inv]
+  exact zero_lt_one
+
+lemma right_pos_of_pos : ∀ j, 0 < x.moveRight j :=
+  fun j => lt_trans x_pos (Numeric.lt_moveRight x_num j)
+
+/-! ### Normalization of a positive numeric game -/
+
+/-- If `x` is a positive numeric game, then keeping only the positive Left options then inserting
+  0 as a Left option, results in an equal game. This game is called the normalization of `x`.
+-/
+def normalization (x : PGame) : PGame :=
+  match x with
+  | ⟨_, r, L, R⟩ => insertLeft ⟨{i // 0 < L i}, r, fun i => L i, R⟩ 0
+
+lemma num_of_normalization_num : (normalization x).Numeric := by
+  rcases x with ⟨xl, xr, xL, xR⟩
+  simp only [normalization]
+  apply insertLeft_numeric
+  · rw [numeric_def] at x_num ⊢
+    simp only [leftMoves_mk, rightMoves_mk, moveLeft_mk, moveRight_mk, Subtype.forall] at x_num ⊢
+    simp only [x_num, implies_true, and_self]
+  · exact numeric_zero
+  · rw [PGame.zero_le]
+    simp only [rightMoves_mk, moveRight_mk]
+    have : 0 ≤ PGame.mk xl xr xL xR := by exact le_of_lt x_pos
+    rw [PGame.zero_le] at this
+    simp only [rightMoves_mk, moveRight_mk] at this
+    exact this
+
+lemma pos_num_eq_normalization :
+    mk x x_num = mk (normalization x) (num_of_normalization_num x_num x_pos) := by
+  simp only [mk_eq_mk]
+  apply PGame.Equiv.trans  -- we show: x ≈ x.insertLeft 0 ≈ normalization x
+  · symm
+    apply insertLeft_equiv_of_lf (lf_of_lt x_pos)
+  rw [equiv_def]
+  rcases x with ⟨xl, xr, xL, xR⟩
+  simp only [normalization, insertLeft]
+  constructor <;> (
+    rw [le_def]
+    simp only [leftMoves_mk, moveLeft_mk, Sum.exists, Sum.elim_inl, Subtype.exists, exists_prop,
+      Sum.elim_inr, exists_const, Sum.forall, le_refl, or_true, zero_rightMoves, IsEmpty.exists_iff,
+      or_false, implies_true, and_true, rightMoves_mk, moveRight_mk]
+  )
+  · constructor
+    · intro i
+      left
+      by_contra bad
+      simp only [not_or, not_exists, not_and, PGame.not_le] at bad
+      rw [lf_iff_lt numeric_zero (x_num.2.1 i)] at bad
+      rcases bad with ⟨bad1, bad2⟩
+      specialize bad1 i bad2
+      exact lf_irrefl _ bad1
+    · intro j
+      right
+      use j
+  · constructor
+    · intro i
+      left
+      left
+      use i
+    · intro j
+      right
+      use j
```

Comment:
```suggestion
lemma eq_normalization_of_pos:
    mk x x_num = mk (normalization x) (num_of_normalization_num x_num x_pos) := by
  rw [mk_eq_mk]
   -- we show: x ≈ x.insertLeft 0 ≈ normalization x
  apply trans <| symm <| insertLeft_equiv_of_lf x_pos.lf
  rw [equiv_def]
  rcases x with ⟨xl, xr, xL, xR⟩
  simp only [normalization, insertLeft]
  constructor <;> (
    rw [le_def]
    simp only [leftMoves_mk, moveLeft_mk, Sum.exists, Sum.elim_inl, Subtype.exists,
      Sum.elim_inr, exists_const, Sum.forall, le_refl, or_true, IsEmpty.exists_iff,
      or_false, implies_true, and_true, moveRight_mk]
  ) <;>
  constructor <;>
  intro i
  · left
    cases' le_or_gf (xL i) 0 with h h
    · right; exact h
    · left; use i, h.lt numeric_zero (x_num.moveLeft _)
  · right; use i
  · left; left; use i
  · right; use i
```
It might be possible to golf this further, `le_def` is usually strictly worse than `le_iff_forall_lf` or `le_iff_forall_lt`.